labs/core_bound/compiler_intrinsics_4/solution.cpp

```diff
@@ -0,0 +1,209 @@
+  constexpr auto& vec_cmpeq_int64 = _mm256_cmpeq_epi64;
+  constexpr auto vec_cmpgt_double = [](auto a, auto b) { return _mm256_cmp_pd(a, b, _CMP_GT_OQ); };
+  constexpr auto& vec_or_mask = _mm256_or_si256;
+  constexpr auto& vec_movemask = _mm256_movemask_pd;
+  #else
+  using Vec = __m128d;
+  constexpr auto& vec_setzero = _mm_setzero_pd;
+  constexpr auto& vec_set1 = _mm_set1_pd;
+  constexpr auto& vec_set1_int = _mm_set1_epi64x;
+  constexpr auto& vec_load = _mm_loadu_pd;
+  constexpr auto vec_store_int = [](auto* a, auto b) { _mm_storeu_si128((__m128i*)a, b); };
+  constexpr auto& vec_add = _mm_add_pd;
+  constexpr auto& vec_add_int = _mm_add_epi64;
+  constexpr auto& vec_sub = _mm_sub_pd;
+  constexpr auto& vec_mul = _mm_mul_pd;
+  constexpr auto& vec_blend = _mm_blendv_pd;
+  constexpr auto& vec_cmpeq_int64 = _mm_cmpeq_epi64;
+  constexpr auto& vec_cmpgt_double = _mm_cmpgt_pd;
+  constexpr auto& vec_or_mask = _mm_or_si128;
+  constexpr auto& vec_movemask = _mm_movemask_pd;
+  #endif
+  #else
+  using Vec = float64x2_t;
+  constexpr auto vec_setzero = []() { return vdupq_n_f64(0.0); };
+  constexpr auto& vec_set1 = vdupq_n_f64;
+  constexpr auto& vec_set1_int = vdupq_n_u64;
+  constexpr auto vec_load = [](const auto* a) { return vld1q_f64(a); };
+  constexpr auto vec_store_int = [](auto* a, auto b) { vst1q_u64(a, b); };
+  constexpr auto& vec_add = vaddq_f64;
+  constexpr auto& vec_add_int = vaddq_u64;
+  constexpr auto& vec_sub = vsubq_f64;
+  constexpr auto& vec_mul = vmulq_f64;
+  constexpr auto vec_blend = [](auto a, auto b, auto m) { return vbslq_f64(m, b, a); };
+  constexpr auto& vec_cmpeq_int64 = vceqq_u64;
+  constexpr auto& vec_cmpgt_double = vcgtq_f64;
+  constexpr auto& vec_or_mask = vorrq_u64;
+  constexpr auto vec_movemask = [](auto m) {
+    const auto bit0 = vgetq_lane_u64(m, 0) & 0b01;
+    const auto bit1 = vgetq_lane_u64(m, 1) & 0b10;
+    return bit1 | bit0;
+  };
+  #endif  // __x86_64__
+  constexpr auto kVecSize = sizeof(Vec) / sizeof(double);
+}  // namespace
+
+//#define SOLUTION
+#ifdef SOLUTION
+
+std::vector<short> mandelbrot(int image_width, int image_height) {
+  const auto data_width = image_width + 2;
+  const auto data_height = image_height + 2;
+  const auto diameter_y = kDiameterX / image_width * image_height;
+  const auto min_x = kCenterX - kDiameterX / 2;
+  const auto max_x = kCenterX + kDiameterX / 2;
+  const auto min_y = kCenterY - diameter_y / 2;
+  const auto max_y = kCenterY + diameter_y / 2;  
+  const size_t data_size = data_width * data_height;
+  std::vector<short> data(data_size);
+  const auto squared_bound = vec_set1(kSquareBound);
+  const auto max_iter = vec_set1_int(kMaxIterations);
+  const auto iter_inc = vec_set1_int(1);
+  auto px = 0;
+  auto py = 0;
+  std::array<double, kVecSize> c_x_arr;
+  std::array<double, kVecSize> c_y_arr;
+  std::array<size_t, kVecSize> res_idx;
+  size_t data_idx = 0;
+  size_t res_used = 0;
+  auto next_data_item = [&](int idx) {
```

Comment:
Can you please add some comments in this function? The more the better :)
This will live in the `golden` branch.